lib/ansible/modules/cloud/vmware/vmware_guest.py

```diff
@@ -1323,29 +1323,47 @@ def get_vm_scsi_controller(self, vm_obj):
 
     def get_configured_disk_size(self, expected_disk_spec):
         # what size is it?
+        disk_size_parse_failed = False
+        expected = None
+        unit = None
         if [x for x in expected_disk_spec.keys() if x.startswith('size_') or x == 'size']:
-            # size_tb, size_gb, size_mb, size_kb, size_b ...?
+            # size, size_tb, size_gb, size_mb, size_kb
             if 'size' in expected_disk_spec:
-                expected = ''.join(c for c in expected_disk_spec['size'] if c.isdigit())
-                unit = expected_disk_spec['size'].replace(expected, '').lower()
-                expected = int(expected)
+                size_regex = re.compile(r'(\d+(?:\.\d+)?)([tgmkTGMK][bB])')
+                disk_size_m = size_regex.match(expected_disk_spec['size'])
+                if disk_size_m:
+                    expected = disk_size_m.group(1)
+                    unit = disk_size_m.group(2)
+                else:
+                    disk_size_parse_failed = True
+                try:
+                    if re.match(r'\d+\.\d+', expected):
+                        # We found float value in string, let's typecast it
+                        expected = float(expected)
+                    else:
+                        # We found int value in string, let's typecast it
+                        expected = int(expected)
+                except (TypeError, ValueError, NameError):
+                    disk_size_parse_failed = True
+
+                if disk_size_parse_failed or not expected or not unit:
+                    # Common failure
+                    self.module.fail_json(msg="Failed to parse disk size please review value"
+                                              " provided using documentation.")
```

Comment:
I propose:
```
if [x for x in expected_disk_spec.keys() if x.startswith('size_') or x == 'size']:
    # size, size_tb, size_gb, size_mb, size_kb
    if 'size' in expected_disk_spec:
        size_regex = re.compile(r'(\d+(?:\.\d+)?)([tgmkTGMK][bB])')
        disk_size_m = size_regex.match(expected_disk_spec['size'])
        try:
            if disk_size_m:
                expected = disk_size_m.group(1)
                unit = disk_size_m.group(2)
            else:
	    	raise ValueError

            if re.match(r'\d+\.\d+', expected):
                # We found float value in string, let's typecast it
                expected = float(expected)
            else:
                # We found int value in string, let's typecast it
                expected = int(expected)

            if not expected or not unit:
	    	raise ValueError
		
        except (TypeError, ValueError, NameError):
            # Common failure
            self.module.fail_json(msg="Failed to parse disk size please review value"
                                      " provided using documentation.")
```

No need to add the following:
```
+        disk_size_parse_failed = False
+        expected = None
+        unit = None
```

@Akasurde What do you think ?

Reply:
Done.